internal/privatestate/data.go

```diff
@@ -0,0 +1,163 @@
+package privatestate
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"unicode/utf8"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+)
+
+// Data contains private state data for the framework and providers.
+type Data struct {
+	// Potential future usage:
+	// Framework contains private state data for framework usage.
+	Framework map[string][]byte
+
+	// Provider contains private state data for provider usage.
+	Provider ProviderData
+}
+
+// Bytes returns a JSON encoded slice of bytes containing the merged
+// framework and provider private state data.
+func (d Data) Bytes(_ context.Context) ([]byte, diag.Diagnostics) {
+	var diags diag.Diagnostics
+
+	bytes, err := json.Marshal(d)
```

Comment:
I think we may need to merge the framework and provider maps before marshalling directly here, so we don't wind up with top-level JSON keys of `Framework` and `Provider` (e.g. we should ensure we can roundtrip this data).